### PR TITLE
Add experimental settings type (v2)

### PR DESCRIPTION
**Pull request by rzhao271**
A newer version of #182692, but different enough that I am making a new PR for it.

### CODE REVIEW

**Review by roblourens — DISMISSED** [review]

File: src/vs/workbench/contrib/preferences/common/preferences.ts [hunk 1]
```diff
@@ -88,3 +91,25 @@ export const REQUIRE_TRUSTED_WORKSPACE_SETTING_TAG = 'requireTrustedWorkspace';
 export const KEYBOARD_LAYOUT_OPEN_PICKER = 'workbench.action.openKeyboardLayoutPicker';
 
 export const ENABLE_LANGUAGE_FILTER = true;
+
+export const ENABLE_EXTENSION_TOGGLE_SETTINGS = true;
```
[the remark this answers is not in the document]
Comment (rzhao271) [reply src/vs/workbench/contrib/preferences/common/preferences.ts:95]: It is a feature flag. One can turn it off to completely turn off the feature.

**Comment by rzhao271** [comment]
Remaining items
- [ ] Get the setting to show up when searched. The PR doesn't affect the config layer anymore, so the setting is not yet registered to show up during searches.
- [ ] Use extensionGalleryService to get the group title.

File: src/vs/workbench/contrib/preferences/browser/settingsEditor2.ts [hunk 2]
```diff
@@ -1210,10 +1243,47 @@ export class SettingsEditor2 extends EditorPane {
 			this.hasWarnedMissingSettings = true;
 		}
 
-		const commonlyUsed = resolveSettingsTree(commonlyUsedData, dividedGroups.core, this.logService);
-		resolvedSettingsRoot.children!.unshift(commonlyUsed.tree);
+		const toggleData = await getExperimentalExtensionToggleData(this.workbenchAssignmentService, this.environmentService, this.productService);
+		if (toggleData) {
+			for (const key in toggleData.settingsEditorRecommendedExtensions) {
+				const prerelease = toggleData.settingsEditorRecommendedExtensions[key].onSettingsEditorOpen!.prerelease;
+				const groupTitle = toggleData.settingsEditorRecommendedExtensions[key].onSettingsEditorOpen!.groupTitle;
```
Comment (sandy081) [inline src/vs/workbench/contrib/preferences/browser/settingsEditor2.ts:1250]: You can get the group title by getting the manifest of the extension as follows:

```ts
const manifest = await this.extensionGalleryService.getManifest(extension, CancellationToken.None);
if (!manifest) {
	continue;
}
groupTitle = manifest.contributes.configuration.title;
```

File: src/vs/workbench/contrib/preferences/browser/settingsEditor2.ts [hunk 3]
```diff
@@ -1210,10 +1243,47 @@ export class SettingsEditor2 extends EditorPane {
...
+					range: nullRange,
+					key: settingKey,
+					keyRange: nullRange,
+					value: null,
+					valueRange: nullRange,
+					description: [extension?.description || ''],
+					descriptionIsMarkdown: false,
+					descriptionRanges: [],
+					title: localize('manageExtension', "Manage {0}", extensionName),
+					scope: ConfigurationScope.APPLICATION,
```
Comment (sandy081) [inline src/vs/workbench/contrib/preferences/browser/settingsEditor2.ts:1272]: May be scope can be made window? Making this application scope shows that it is applied for All profiles and I think it is irrelevant here.

File: src/vs/workbench/contrib/preferences/browser/settingsTree.ts [hunk 4]
```diff
@@ -1878,6 +1890,66 @@ export class SettingBoolRenderer extends AbstractSettingRenderer implements ITre
...
+
+		return template;
+	}
+
+	renderElement(element: ITreeNode<SettingsTreeSettingElement, never>, index: number, templateData: ISettingExtensionToggleItemTemplate): void {
+		super.renderSettingElement(element, index, templateData);
+	}
+
+	renderButton(extensionId: string, actionButton: Button): void {
+		this._extensionsService.getExtension(extensionId).then(extension => {
```
Comment (sandy081) [inline src/vs/workbench/contrib/preferences/browser/settingsTree.ts:1927]: I supposed you do not need to get the name from extension service any more. I presume you obtained the name of the extension already from gallery service.

File: src/vs/workbench/contrib/preferences/browser/settingsTree.ts [hunk 5]
```diff
@@ -1878,6 +1890,66 @@ export class SettingBoolRenderer extends AbstractSettingRenderer implements ITre
...
+	protected renderValue(dataElement: SettingsTreeSettingElement, template: ISettingExtensionToggleItemTemplate, onChange: (_: undefined) => void): void {
+		template.elementDisposables.clear();
+
+		const extensionId = dataElement.setting.extensionId!;
+		template.elementDisposables.add(template.actionButton.onDidClick(async () => {
+			this._telemetryService.publicLog2<{ extensionId: String }, ManageExtensionClickTelemetryClassification>('ManageExtensionClick', { extensionId });
+			this._commandService.executeCommand('extension.open', extensionId);
+		}));
+
+		template.elementDisposables.add(this._extensionsWorkbenchService.onChange((e) => {
```
Comment (sandy081) [inline src/vs/workbench/contrib/preferences/browser/settingsTree.ts:1943]: I also think this listener is not needed anymore

**Review by sandy081 — COMMENTED** [review]
Provided feedback

File: src/vs/workbench/contrib/preferences/browser/settingsTree.ts (hunk 4 above)
Comment (rzhao271) [reply src/vs/workbench/contrib/preferences/browser/settingsTree.ts]: True. I'll keep the button saying "Manage extension" though, to make it clear that it's an extension.

File: src/vs/workbench/contrib/preferences/browser/settingsEditor2.ts [hunk 6]
```diff
@@ -1210,10 +1251,62 @@ export class SettingsEditor2 extends EditorPane {
...
 
-		const commonlyUsed = resolveSettingsTree(commonlyUsedData, dividedGroups.core, this.logService);
+		const additionalGroups: ISettingsGroup[] = [];
+		const toggleData = await getExperimentalExtensionToggleData(this.workbenchAssignmentService, this.environmentService, this.productService);
+		if (toggleData && groups.filter(g => g.extensionInfo).length) {
+			for (const key in toggleData.settingsEditorRecommendedExtensions) {
+				const prerelease = toggleData.settingsEditorRecommendedExtensions[key].onSettingsEditorOpen!.prerelease;
+
+				const extensionId = (typeof prerelease === 'string' && this.productService.quality !== 'stable') ? prerelease : key;
+				const galleryExtensions = await this.extensionGalleryService.getExtensions([{ id: extensionId }], CancellationToken.None);
```
Comment (sandy081) [inline src/vs/workbench/contrib/preferences/browser/settingsEditor2.ts:1261]: You can directly assign extension

```
const [extension] = await this.extensionGalleryService.getExtensions([{ id: extensionId }], CancellationToken.None);
```

**Review by sandy081 — DISMISSED** [review]

**Review by sandy081 — APPROVED** [review]

**Pull request merged**